### PR TITLE
Add HTTPRequestWrappingIRequest

### PR DESCRIPTION
**Pull request by wsanchez**

### CODE REVIEW

**Comment by wsanchez** [comment]
I'm confused about the [coverage report](https://codecov.io/gh/twisted/klein/compare/ca890d1ceb744908fa76d7eada44904c1104ee07...0c84c65592a3e7ee1420476d1478874357fcf7e3/diff) here, which indicates that `http_urls` [is not called](https://codecov.io/gh/twisted/klein/compare/ca890d1ceb744908fa76d7eada44904c1104ee07...0c84c65592a3e7ee1420476d1478874357fcf7e3/diff#D1-219) but also [that it is](https://codecov.io/gh/twisted/klein/compare/ca890d1ceb744908fa76d7eada44904c1104ee07...0c84c65592a3e7ee1420476d1478874357fcf7e3/diff#D4-80).

That may be related to [this ticket](https://github.com/python-attrs/attrs/issues/317) in attrs.

File: src/klein/test/_strategies.py [hunk 1]
```diff
@@ -26,10 +35,56 @@
...
+    """
+    global _idnaCharacters
+
+    if _idnaCharacters is None:
+        result = []
+
+        # Data source "IDNA Derived Properties":
+        # https://www.iana.org/assignments/idna-tables-6.3.0/
+        #   idna-tables-6.3.0.xhtml#idna-tables-properties
+        dataFileName = join(dirname(__file__), "idna-tables-properties.csv")
```
Comment (moshez) [inline src/klein/test/_strategies.py]: The right way to access data files is with `pkg_resources`. See [the docs](http://setuptools.readthedocs.io/en/latest/pkg_resources.html#basic-resource-access).

File: src/klein/test/_strategies.py [hunk 2]
```diff
@@ -61,3 +116,125 @@ def latin1_text(draw, min_size=None, max_size=None):
...
+    """
+    return draw(
+        iterables(
+            text(min_size=1, alphabet=_path_characters),
+            max_size=10, average_size=3,
+        )
+    )
+
+
+@composite
```
Comment (moshez) [inline src/klein/test/_strategies.py]: This seems like something that would make sense to add to the `hyperlink` package. If this is for expedience, so as not to wait for a hyperlink release, this makes sense.

File: src/klein/test/_trial.py [hunk 3]
```diff
@@ -28,18 +28,16 @@ class TestCase(_TestCase):
 
     if (twistedVersion.major, twistedVersion.minor) < (16, 4):
         def assertRegex(self, text, regex, msg=None):
+            # type: (str, Any, str) -> None
```
Comment (moshez) [inline src/klein/test/_trial.py]: This seems like an unrelated change?

File: src/klein/_request_compat.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        else:
+            default = 80
+        if port != default:
+            netloc += u":{}".format(port)
+
+        path = nativeString(request.uri)
+        if path and path[0] == u"/":
+            path = path[1:]
+
+        return URL.fromText(u"{}://{}/{}".format(scheme, netloc, path))
```
[the remark this answers is not in the document]
Comment (wsanchez) [reply src/klein/_request_compat.py]: This is significantly complicated by the fact that `path` and `host` may be using either bytes or text (docs say both are `str` but in practice I think they are really `AnyStr`).

File: src/klein/test/_strategies.py (hunk 1 above)
Comment (wsanchez) [reply src/klein/test/_strategies.py]: `csv.reader` wants a text stream and `resource_stream` only provides a binary stream.  Not a problem! I'll wrap it in `TextIOWrapper`, which works in Py3 but not in Py2, because `file`, which is what you get from `resource_stream`, does not implement the `IOBase` interface for who the hell knows what reasons.

Comment (wsanchez) [reply src/klein/test/_strategies.py]: I could read the whole file an shove it into a `TextIO`… but it's not a tiny file so that feels icky.

File: src/klein/test/_trial.py (hunk 3 above)
Comment (wsanchez) [reply src/klein/test/_trial.py]: yeah, it's opportunistic cleanup :-/

File: src/klein/test/_strategies.py (hunk 2 above)
Comment (wsanchez) [reply src/klein/test/_strategies.py]: yes, agreed

File: src/klein/_request_compat.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    @property
+    def headers(self):
+        # type: () -> IHTTPHeaders
+        return HTTPHeadersWrappingHeaders(headers=self._request.requestHeaders)
+
+
+    def bodyAsFount(self):
+        # type: () -> IFount
+        source = self._request.content
```
[the remark this answers is not in the document]
Comment (wsanchez) [reply src/klein/_request_compat.py]: 476673d87db0cebff35f787e025acdcb42cc6672
